### PR TITLE
 ID Refactor

### PR DESCRIPTION
**Pull request by MACIEKBAKI**
## About the Pull Request

id refactor
## Why It's Good For The Game

work actually being done
## Changelog

:cl:
refactor: Complete refactor of our ID system. Now uses defines.
add: New guestpass TGUI
/:cl:

TODO:

- [x] Set up proper access requirements for containment access (i need to go through it or someone to go through it and tell me what scp containment access should be what

### CODE REVIEW

**Comment by MACIEKBAKI** [comment]
might datumize access regions idk

**Comment by TheTitanicEngineer** [comment]
On the current code, access regions are based on defines, and access is based on datums.

I do not believe this PR brings any benefit to the game, making it into datum adds one extra step. If I were to create a new access after this PR, I would need to define the access and make the access in the access.dm file, it's one extra step that I do not see any purpose for it.

**Comment by MACIEKBAKI** [comment]
> On the current code, access regions are based on defines, and access is based on datums.
> 
> I do not believe this PR brings any benefit to the game, making it into datum adds one extra step. If I were to create a new access after this PR, I would need to define the access and make the access in the access.dm file, it's one extra step that I do not see any purpose for it.

making *what* into datums, precisely? I can't quite deduce what you're reffering to

If it's datumizing access regions: seldom do you need to add a new access region and it would make some things (for instance, sorting through all regions and displaying stuff per-region, maybe?) easier for, say, the ID console
If it's datumizing access itself: Current-code access *already is a datum*. I'm just making it use `#define`s because global constant variables are pretty ugly, adding a new access with this pr will not be that different to adding a new access without it, except that you have to change two files instead of just one

**Review by Tsurupeta — APPROVED** [review]
https://hackmd.io/@f19/access

We'll need to implement that design doc later

File: code/__defines/access_defs.dm [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+#define ACCESS_REGION_ENGINEERING   4
+#define ACCESS_REGION_COMMAND       5
+#define ACCESS_REGION_GENERAL       6
+#define ACCESS_REGION_SUPPLY        7
+#define ACCESS_REGION_SERVICE       8
+
+#define ACCESS_REGION_CONTAINMENT	9
+#define ACCESS_REGION_OFFSITE		10
+
+// Keep those two up to date
```
Comment (Penwin0) [inline code/__defines/access_defs.dm:15]: This was not kept up to date

File: code/__defines/access_defs.dm [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+#define ACCESS_COM_COMMS "ACCESS_COMMS_COMMAND"
+
+#define ACCESS_SECURITY "ACCESS_SECURITY" //1
+#define ACCESS_BRIG "ACCESS_BRIG" // Brig timers and permabrig 2
+#define ACCESS_ARMORY "ACCESS_ARMORY" //3
+#define ACCESS_FORENSICS_LOCKERS "ACCESS_FORENSICS" //4
+#define ACCESS_MEDICAL "ACCESS_MEDICAL" //5
+#define ACCESS_MORGUE "ACCESS_MORGUE" //6
+#define ACCESS_TOX "ACCESS_TOXINS" //7
+#define ACCESS_TOX_STORAGE "ACCESS_TOX_STORAGE" //8
```
Comment (Penwin0) [inline code/__defines/access_defs.dm:125]: Are you certain these numbers are correct?

File: code/game/jobs/access/_access.dm [hunk 3]
```diff
@@ -134,6 +148,16 @@
 
 	return priv_station_access.Copy()
 
+/var/list/priv_innate_access
```
Comment (Penwin0) [inline code/game/jobs/access/_access.dm:151]: ¿?

File: code/game/jobs/access/access_command.dm [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	id = ACCESS_CAPTAIN
+	desc = "Captain"
+	region = ACCESS_REGION_COMMAND
+
+/datum/access/all_personal_lockers
+	id = ACCESS_ALL_PERSONAL_LOCKERS
+	desc = "Personal Lockers"
+	region = ACCESS_REGION_COMMAND
+
+/datum/access/lawyer
```
Comment (Penwin0) [inline code/game/jobs/access/access_command.dm:70]: Ah yes, command

File: code/game/jobs/job/civilian.dm [hunk 5]
```diff
@@ -8,8 +8,8 @@
 	spawn_positions = 1
 	supervisors = "the head of personnel"
 	selection_color = "#515151"
-	access = list(access_hydroponics, access_bar, access_kitchen)
```
Comment (Penwin0) [inline code/game/jobs/job/civilian.dm:11]: Might as well destroy all this filthy commented code

File: code/modules/ascent/ascent_atmos.dm [hunk 6]
```diff
@@ -13,11 +13,11 @@
 MANTIDIFY(/obj/machinery/alarm, "mantid thermostat", "atmospherics")
 
 /obj/machinery/alarm/ascent
-	req_access = list(access_ascent)
```
Comment (Penwin0) [inline code/modules/ascent/ascent_atmos.dm:16]: Kill these files honestly

File: maps/site53/job/access/access_containment.dm [hunk 7]
```diff
@@ -0,0 +1,204 @@
+/datum/access/containment_scp_012
```
Comment (Penwin0) [inline maps/site53/job/access/access_containment.dm:1]: I hate how this stuff is all under maps, but not a fault of this PR

File: code/__defines/access_defs.dm (hunk 2 above)
Comment (MACIEKBAKI) [reply code/__defines/access_defs.dm:125]: i dont know why we have them all accesses are strings anyway
i used regex to mass produce the #defines

File: code/__defines/access_defs.dm (hunk 1 above)
Comment (MACIEKBAKI) [reply code/__defines/access_defs.dm:15]: these are used only in two places; airlock electronics and access decrypter to determine regions
containment and offsite dont really fit there so i've decided to not keep it up to date (read: lazy)

File: code/game/jobs/access/_access.dm (hunk 3 above)
Comment (MACIEKBAKI) [reply code/game/jobs/access/_access.dm:151]: access levels like sec level 5, sci lvl 4, med lvl 2, admin lvl 3 etc

**Comment by MACIEKBAKI** [comment]
uhoh

**Pull request merged**